### ISSUE TITLE
There needs to be an easier way for a user to list all current resource limits

### ISSUE THREAD

**Issue by krader1961** [elves/elvish#1494]
I understand why @xiaq chose to implement support for manipulating resource limits on UNIX platforms using a variable (`unix:rlimits` similar to `unix:umask`) rather than merge my change to implement a `unix:ulimit` command loosely modeled on the POSIX `ulimit` command (see PR #1474). However, it's not reasonable to expect a user to copy/paste a complicated pipeline to do the equivalent of `ulimit -a` in order to see the current limits. I propose keeping the current `unix:rlimits` variable for manipulating the resource limits but add a `unix:ulimits` command implemented as a builtin Elvish script function that **only** displays all the limits ala `ulimit -a` in a POSIX shell. The output of that function should not only display all limits but also include verbiage explaining how to change a limit.

**Comment by krader1961** [comment]
There could also be a `builtin:ulimit` command that invokes `unix:ulimits` on UNIX like systems and emit a reasonable error message on non-UNIX like systems. That would maximize the friendliness for people migrating from other shells (including non-POSIX shells like Fish). However, that seems like [a bridge too far](https://www.historyextra.com/period/a-bridge-too-far/); that is, needlessly catering to legacy shell users migrating to Elvish.

**Comment by krader1961** [comment]
I started working on implementing this and realized the existing mechanisms for evaluating "builtin" Elvish code in the context of initializing modules is a hodgepodge of approaches. The only idiom for a module that augments its behavior with Elvish script is the `edit` module which does so via a `initElvishState()` function that is run as a byproduct of executing `edit.NewEditor()`. That happens as a side-effect of instantiating an interactive editor session and can't be applied to non-interactive contexts such as augmenting the `unix:` namespace with variables and functions defined by Elvish script. It seems to me that the inclusion of Elvish script in the initialization of builtin modules needs refactoring to provide a unified mechanism for doing so. At the same time `BundledModules` should be replaced by the new unified mechanism even if the native Go package for the Elvish module does nothing more than create a framework for the Elvish script implementation.

**Comment by xiaq** [comment]
I don't think there should be a dedicated command to show the content of `$unix:rlimits`.

Elvish doesn't have dedicated commands for showing the content of `$edit:insert:binding`, `$edit:abbr`, or `$edit:completion:arg-completer`; you just use `pprint`.

What should be is to make `pprint` a bit smarter. This is the current output of `pprint $unix:rlimits` on my machine:

```
[
 &nofile=	[
   &cur=	(num 256)
  ]
 &fsize=	[&]
 &nproc=	[
   &max=	(num 2666)
   &cur=	(num 2666)
  ]
 &memlock=	[&]
 &cpu=	[&]
 &core=	[
   &cur=	(num 0)
  ]
 &stack=	[
   &max=	(num 67092480)
   &cur=	(num 8372224)
  ]
 &rss=	[&]
 &data=	[&]
]
```

But ideally it should look like this instead:

```
[
 &cpu=     [&]
 &core=    [&cur=(num 0)]
 &data=    [&]
 &fsize=   [&]
 &memlock= [&]
 &nofile=  [&cur=(num 256)]
 &nproc=   [&cur=(num 2666) &max=(num 2666)]
 &rss=     [&]
 &stack=   [&cur=(num 8372224) &max=(num 67092480)]
]
```

**Comment by krader1961** [comment]
I'm all for improving the output of `pprint`. However, that doesn't address this issue. The output of POSIX `ulimit -a` is meant to be read by a shell **user**. While the output of `pprint` is meant to be used to by a **programmer** (who is probably trying to understand why their code doesn't do what they expect).

Whether there should be dedicated commands for producing more human friendly output of data such as `$edit:abbr`, rather than relying on `pprint`, is an interesting question. Especially since that particular example doesn't also include the data from `$edit:small-word-abbr`. Why should the user have to execute two `pprint` commands to see all the abbreviations in effect? Also, the display of that information is, again, not formatted in a manner friendly to a user of the shell who isn't also a programmer.

**Comment by krader1961** [comment]
Consider what this output of a hypothetical `pprint` outputs:

```
&rss=     [&]
```

What does that mean with regard to the "RSS" resource limit? The empty map will only be meaningful to a programmer who understands how Elvish implements UNIX resource limits. It won't make any sense to an Elvish shell **user**.

**Comment by xiaq** [comment]
The line between "user" and "programmer" is quite fuzzy when it comes to a shell.

I don't deny that a text output optimized for this use case (such as the one your PR implemented) reads better and may be more compelling for some users. Bespoke commands and formats are always going to be more compelling for the particular domain they're optimized for; even with a better `pprint`, bespoke commands to list key bindings, abbreviations and completers can all improve upon the generic `pprint` format. Other shells have bespoke commands to do all of those anyway.

However, my design methodology values uniformity and generalizability over the gains of bespoke commands and formats. Providing `unix:rlimits` as a map of maps does require the user to learn how to interpret the maps. In particular, they have to learn that the absence of `&cur` or `&max` means the absence of the corresponding limits (more subtle than an explicit "unlimited", but hardly surprising).

However, once they learn that, they immediately know how to obtain one resource limit (by indexing the map), how to mutate the map (with `set`), and how to serialize it to JSON (with `to-json`) to perhaps store in a file. This is the power of uniform abstractions that you will lose with bespoke commands.

**Comment by krader1961** [comment]
> The line between "user" and "programmer" is quite fuzzy when it comes to a shell.

I agree. In fact, you can make a reasonable argument that every interactive shell user is a "programmer" if they do anything more basic than executing commands. For example, using the POSIX `cmd1 && cmd2` syntax is writing a program. It should be possible for users to use low level primitives (e.g., `$unix:rlimits`) to display that information in a manner they prefer. That's why I acknowledged that exposing that information via a variable (`unix:rlimits`) rather than a command (`unix:ulimit`) is preferable. That doesn't preclude providing useful predefined commands for common operations such as listing the current resource limits or interactive command abbreviations. The user can always chose to display the information in a different fashion given the available primitives.

**Comment by xiaq** [comment]
> That doesn't preclude providing useful predefined commands for common operations such as listing the current resource limits or interactive command abbreviations. The user can always choose to display the information in a different fashion given the available primitives.

It doesn't and you can certainly write Elvish modules for those purposes. But at the moment I don't see a compelling case to include them in Elvish itself.

**Comment by krader1961** [comment]
FWIW, I ultimately wrote a *~/.config/elvish/lib/util_unix.elv* module that I use like this (from my *rc.elv*):

```
if $platform:is-unix {
  use unix
  use util_unix
  edit:add-var unix: $unix:
  edit:add-var ulimit~ $util_unix:ulimit~
}
```

The contents of *util_unix.elv*:

```
use unix

fn -ulimit-hdr {||
        printf "%-10s %10s %10s\n" "resource" "current" "maximum"
        printf "%-10s %10s %10s\n" "==========" "==========" "=========="
}

fn -ulimit-print {|resource vals|
    var cur max = unlimited unlimited
    if (has-key $vals cur) {
        set cur = $vals[cur]
    }
    if (has-key $vals max) {
        set cur = $vals[cur]
    }
    printf "%-10s %10v %10v\n" $resource $cur $max
}

# Usage:
#   ulimit
#      Display all resource limits.
#   ulimit $resource
#      Display one resource limit.
#   ulimit $resource $value
#      Update a resource limit. The value can be an integer, or the strings
#      "inf" or "unlimited".
fn ulimit {|@args|
    if (== 0 (count $args)) {
        # Display all resource limits.
        -ulimit-hdr
        for resource [(keys $unix:rlimits | order)] {
            -ulimit-print $resource $unix:rlimits[$resource]
        }
    } elif (== 1 (count $args)) {
        # Display one resource limit.
        var resource = $args[0]
        if (has-key $unix:rlimits $resource) {
            -ulimit-hdr
            -ulimit-print $resource $unix:rlimits[$resource]
        } else {
            fail "no such resource: "$resource
        }
    } elif (== 2 (count $args)) {
        # Modify the current limit for a resource.
        var resource value = $args[0] $args[1]
        if (not (has-key $unix:rlimits $resource)) {
            fail "no such resource: "$resource
        }
        if (or (eq $value "inf") (eq $value "unlimited")) {
            del unix:rlimits[$resource][cur]
        } else {
            try {
                set value = (num $value)
            } catch {
                fail $resource" value "$value" is not a number"
            }
            set unix:rlimits[$resource][cur] = $value
        }
    } else {
        fail "expected at most two arguments, got "(count $args)
    }
}
```

Yes, I can (and probably should) publish this on [Awesome Elvish](https://github.com/elves/awesome-elvish), but I still feel that this is so fundamental to using Elvish on UNIX it shouldn't require installing a third-party module. Imagine Bash, Zsh, Ksh, or any other legacy UNIX shell that required the user to write their own `ulimit` function to get this functionality.

**Comment by krader1961** [comment]
For the benefit of anyone stumbling upon this issue I'll note that I've published a module that provides a human friendly command for displaying and manipulating UNIX resource limits. See https://github.com/krader1961/elvish-lib/ (which has been added to the [Awesome Elvish](https://github.com/elves/awesome-elvish) page. For example, on my primary macOS server:

```
> ulimit
resource      current    maximum
========== ========== ==========
core                0  unlimited
cpu         unlimited  unlimited
data        unlimited  unlimited
fsize       unlimited  unlimited
memlock     unlimited  unlimited
nofile            256  unlimited
nproc            5333  unlimited
rss         unlimited  unlimited
stack         8372224  unlimited
```

I'm still inclined to think this feature should be included by default in the Elvish program. Nonetheless, I am closing this issue since it is easy enough for anyone to deal with resource limits interactively in a more human friendly manner by using my module, or one written by someone else.
